pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go

```diff
@@ -268,7 +241,74 @@ func getHostBasedServiceUrl(isvc *v1beta1.InferenceService, config *v1beta1.Ingr
 	}
 }
 
-func (r *IngressReconciler) reconcileExternalService(ctx context.Context, isvc *v1beta1.InferenceService, config *v1beta1.IngressConfig) error {
+func (ir *IngressReconciler) reconcileVirtualService(ctx context.Context, isvc *v1beta1.InferenceService) error {
+	disableIstioVirtualHost := ir.ingressConfig.DisableIstioVirtualHost
+
+	// Check if existing knative service name has default suffix
+	defaultNameExisting := &knservingv1.Service{}
+	useDefault := false
+	err := ir.client.Get(ctx, types.NamespacedName{Name: constants.DefaultPredictorServiceName(isvc.Name), Namespace: isvc.Namespace}, defaultNameExisting)
+	if err == nil {
+		useDefault = true
+	}
+
+	domainList := getDomainList(ctx, ir.clientset)
+	desiredIngress := createIngress(isvc, useDefault, ir.ingressConfig, domainList, ir.isvcConfig) // actually the virtual service
+
+	existing := &istioclientv1beta1.VirtualService{}
+	getExistingErr := ir.client.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace}, existing)
+
+	if !isvc.GetForceStopRuntime() {
+		// When Istio virtual host is disabled, we return the underlying component url.
+		// When Istio virtual host is enabled. we return the url using inference service virtual host name and redirect to the corresponding transformer, predictor or explainer url.
+		if !disableIstioVirtualHost {
+			if desiredIngress == nil {
+				return nil
+			}
+
+			err = controllerutil.SetControllerReference(isvc, desiredIngress, ir.scheme)
+			if err != nil {
+				return errors.Wrapf(err, "fails to set owner reference for ingress")
+			}
+
+			if getExistingErr != nil {
+				if apierr.IsNotFound(getExistingErr) {
+					log.Info("Creating Ingress for isvc", "namespace", desiredIngress.Namespace, "name", desiredIngress.Name)
+					err = ir.client.Create(ctx, desiredIngress)
+				}
+			} else {
+				if !routeSemanticEquals(desiredIngress, existing) {
+					deepCopy := existing.DeepCopy()
+					deepCopy.Spec = *desiredIngress.Spec.DeepCopy()
+					deepCopy.Annotations = desiredIngress.Annotations
+					deepCopy.Labels = desiredIngress.Labels
+					log.Info("Update Ingress for isvc", "namespace", desiredIngress.Namespace, "name", desiredIngress.Name)
+					err = ir.client.Update(ctx, deepCopy)
+				}
+			}
+
+			if err != nil {
+				return errors.Wrapf(err, "fails to create or update ingress")
+			}
+		}
+	} else {
+		// Delete the virtualservice
+		if getExistingErr == nil {
+			log.Info("The InferenceService ", isvc.Name, " is marked as stopped — delete its associated VirtualService")
+			if err := ir.client.Delete(ctx, existing); err != nil {
```

Comment:
Following up on https://github.com/kserve/kserve/pull/4337#discussion_r2062391117:

The very first statement in the func is `disableIstioVirtualHost := ir.ingressConfig.DisableIstioVirtualHost`. If it evaluates to `true` no virtual service is created and users may freely create their own.

Here, there is no check of the `disableIstioVirtualHost` nor a check for ownership. Thus, this code can delete a user-owned VirtualService. 

I prefer checking for ownership as that is more accurate than the flag.

Reply:
This same comment applies inside `reconcileExternalService`.